File: app/src/main/java/com/app/ralaunch/renderer/TurnipLoader.java
package com.app.ralaunch.renderer;

import android.system.ErrnoException;
import android.system.Os;
import android.util.Log;
import com.app.ralaunch.utils.AppLogger;

public class TurnipLoader {
    private static final String TAG = "TurnipLoader";
    private static boolean sTurnipLoaded = false;
    private static long sTurnipHandle = 0;
    private static long sVulkanLoaderHandle = 0;
    private static long sVkGetInstanceProcAddr = 0;

    // 确保 libmain.so 已加载（包含 JNI native 方法实现）
    static {
        try {
            System.loadLibrary("main");
            AppLogger.info(TAG, "libmain.so loaded successfully");
        } catch (UnsatisfiedLinkError e) {
            AppLogger.error(TAG, "Failed to load libmain.so: " + e.getMessage());
        }
    }

    /**
     * Load Turnip driver using zomdroid-style linker hook approach
     * @param nativeLibDir The directory containing libvulkan_freedreno.so
     * @param cacheDir The cache directory for SONAME-patched libvulkan.so
     * @return true if Turnip loaded successfully
     */
    public static boolean loadTurnip(String nativeLibDir, String cacheDir) {
        if (sTurnipLoaded) {
            AppLogger.info(TAG, "Turnip already loaded");
            return true;
        }

        try {
            // Set TMPDIR for native code
            if (cacheDir != null) {
                try {
                    Os.setenv("TMPDIR", cacheDir, true);
                    AppLogger.info(TAG, "Set TMPDIR=" + cacheDir);
                } catch (ErrnoException e) {
                    AppLogger.error(TAG, "Failed to set TMPDIR: " + e.getMessage());
                }
            }
            
            // Load Turnip using zomdroid-style approach
            boolean success = nativeLoadTurnip(nativeLibDir, cacheDir);
            if (success) {
                sTurnipHandle = nativeGetTurnipHandle();
                sVulkanLoaderHandle = nativeGetVulkanLoaderHandle();
                sVkGetInstanceProcAddr = nativeGetVkGetInstanceProcAddr();
                sTurnipLoaded = true;
                
                AppLogger.info(TAG, "Turnip loaded successfully (zomdroid-style)!");
                AppLogger.info(TAG, "  Turnip handle: 0x" + Long.toHexString(sTurnipHandle));
                AppLogger.info(TAG, "  Vulkan loader handle: 0x" + Long.toHexString(sVulkanLoaderHandle));
                AppLogger.info(TAG, "  vkGetInstanceProcAddr: 0x" + Long.toHexString(sVkGetInstanceProcAddr));
                
                // Set environment variables for DXVK
                try {
                    Os.setenv("TURNIP_HANDLE", Long.toHexString(sTurnipHandle), true);
                    Os.setenv("VULKAN_PTR", "0x" + Long.toHexString(sVulkanLoaderHandle), true);
                    Os.setenv("VK_GET_INSTANCE_PROC_ADDR", "0x" + Long.toHexString(sVkGetInstanceProcAddr), true);
                } catch (ErrnoException e) {
                    AppLogger.error(TAG, "Failed to set environment variables: " + e.getMessage());
                }
            } else {
                AppLogger.error(TAG, "Failed to load Turnip");
            }
            return success;
        } catch (UnsatisfiedLinkError e) {
            AppLogger.error(TAG, "Native method not found: " + e.getMessage());
            return false;
        }
    }

    public static boolean isTurnipLoaded() {
        return sTurnipLoaded;
    }

    public static long getTurnipHandle() {
        return sTurnipHandle;
    }

    public static long getVulkanLoaderHandle() {
        return sVulkanLoaderHandle;
    }

    public static long getVkGetInstanceProcAddr() {
        return sVkGetInstanceProcAddr;
    }

    // Native methods - implemented in turnip_loader_nsbypass.cpp
    private static native boolean nativeLoadTurnip(String nativeLibDir, String cacheDir);
    private static native long nativeGetTurnipHandle();
    private static native long nativeGetVulkanLoaderHandle();
    private static native long nativeGetVkGetInstanceProcAddr();
}

File: app/src/main/java/com/app/ralaunch/renderer/TurnipLoader.kt
package com.app.ralaunch.renderer

import android.system.ErrnoException
import android.system.Os
import com.app.ralaunch.utils.AppLogger

/**
 * Turnip 驱动加载器
 */
object TurnipLoader {
    private const val TAG = "TurnipLoader"
    
    @JvmStatic
    var isTurnipLoaded: Boolean = false
        private set

    @JvmStatic
    var turnipHandle: Long = 0
        private set

    @JvmStatic
    var vulkanLoaderHandle: Long = 0
        private set

    @JvmStatic
    var vkGetInstanceProcAddr: Long = 0
        private set

    init {
        try {
            System.loadLibrary("main")
            AppLogger.info(TAG, "libmain.so loaded successfully")
        } catch (e: UnsatisfiedLinkError) {
            AppLogger.error(TAG, "Failed to load libmain.so: ${e.message}")
        }
    }

    @JvmStatic
    fun loadTurnip(nativeLibDir: String, cacheDir: String?): Boolean {
        if (isTurnipLoaded) {
            AppLogger.info(TAG, "Turnip already loaded")
            return true
        }

        return try {
            cacheDir?.let {
                try {
                    Os.setenv("TMPDIR", it, true)
                    AppLogger.info(TAG, "Set TMPDIR=$it")
                } catch (e: ErrnoException) {
                    AppLogger.error(TAG, "Failed to set TMPDIR: ${e.message}")
                }
            }

            val success = nativeLoadTurnip(nativeLibDir, cacheDir)
            if (success) {
                turnipHandle = nativeGetTurnipHandle()
                vulkanLoaderHandle = nativeGetVulkanLoaderHandle()
                vkGetInstanceProcAddr = nativeGetVkGetInstanceProcAddr()
                isTurnipLoaded = true

                AppLogger.info(TAG, "Turnip loaded successfully (zomdroid-style)!")
                AppLogger.info(TAG, "  Turnip handle: 0x${turnipHandle.toString(16)}")
                AppLogger.info(TAG, "  Vulkan loader handle: 0x${vulkanLoaderHandle.toString(16)}")
                AppLogger.info(TAG, "  vkGetInstanceProcAddr: 0x${vkGetInstanceProcAddr.toString(16)}")

                try {
                    Os.setenv("TURNIP_HANDLE", turnipHandle.toString(16), true)
                    Os.setenv("VULKAN_PTR", "0x${vulkanLoaderHandle.toString(16)}", true)
                    Os.setenv("VK_GET_INSTANCE_PROC_ADDR", "0x${vkGetInstanceProcAddr.toString(16)}", true)
                } catch (e: ErrnoException) {
                    AppLogger.error(TAG, "Failed to set environment variables: ${e.message}")
                }
            } else {
                AppLogger.error(TAG, "Failed to load Turnip")
            }
            success
        } catch (e: UnsatisfiedLinkError) {
            AppLogger.error(TAG, "Native method not found: ${e.message}")
            false
        }
    }

    private external fun nativeLoadTurnip(nativeLibDir: String, cacheDir: String?): Boolean
    private external fun nativeGetTurnipHandle(): Long
    private external fun nativeGetVulkanLoaderHandle(): Long
    private external fun nativeGetVkGetInstanceProcAddr(): Long
}
